coverage_comment/coverage.py

```diff
@@ -211,57 +217,94 @@ def extract_info(data: dict, coverage_path: pathlib.Path) -> Coverage:
     )
 
 
-def get_diff_coverage_info(base_ref: str, coverage_path: pathlib.Path) -> DiffCoverage:
-    subprocess.run("git", "fetch", "--depth=1000", path=pathlib.Path.cwd())
-    subprocess.run("coverage", "xml", path=coverage_path)
-    with tempfile.NamedTemporaryFile("r") as f:
-        subprocess.run(
-            "diff-cover",
-            "coverage.xml",
-            f"--compare-branch=origin/{base_ref}",
-            f"--json-report={f.name}",
-            "--diff-range-notation=..",
-            "--quiet",
-            path=coverage_path,
+def get_diff_coverage_info(
+    added_lines: dict[pathlib.Path, list[int]], coverage: Coverage
+) -> DiffCoverage:
+    files = {}
+    total_num_lines = 0
+    total_num_violations = 0
+    num_changed_lines = 0
+
+    for path, added_lines_for_file in added_lines.items():
+        num_changed_lines += len(added_lines_for_file)
+
+        try:
+            file = coverage.files[path]
+        except KeyError:
+            continue
+
+        executed = set(file.executed_lines) & set(added_lines_for_file)
+        count_executed = len(executed)
+
+        missing = set(file.missing_lines) & set(added_lines_for_file)
+        count_missing = len(missing)
+        # We don't want to count twice the lines that are both executed and
+        # missing (e.g. the branch lines partially executed)
+        count_total = len(executed | missing)
+
+        total_num_lines += count_total
+        total_num_violations += count_missing
+
+        percent_covered = decimal.Decimal("1")
+        if count_total != 0:
+            # You can't multiply a decimal by a float but you can divide it
+            # by an int, that's why we split the following into 2 lines
+            percent_covered *= count_executed
+            percent_covered /= count_total
```

Comment:
If it is indeed the case that there are lines that occur in both `executed` and `missing`, then we have to change the calculation system here. For example, if `executed` and `missing` contained exactly the same lines, because they are branches that are always only half executed, then something around 50% should come out. But since `count_total` only takes each line into account once, the result would be 100%.

Reply:
Thankfully, you were right and lines only appear once.